### PR TITLE
[16.01] Encoding in libraries

### PR DESCRIPTION
**Pull request by dannon**
WIP, fixes handling of non-ascii chars in library names, file paths, more to come.  Spurred on by -dev email.

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/remote_files.py [hunk 1]
```diff
@@ -132,20 +132,18 @@ def __create_jstree( self, directory, disable='folders' ):
...
                     dir_path = os.path.relpath( os.path.join( dirpath, dirname ), directory )
-                    dir_path_hash = hashlib.sha1( dir_path.encode('utf-8') ).hexdigest()
+                    dir_path_hash = hashlib.sha1(dir_path).hexdigest()
                     disabled = True if disable == 'folders' else False
                     jstree_paths.append( jstree.Path( dir_path, dir_path_hash, { 'type': 'folder', 'state': { 'disabled': disabled }, 'li_attr': { 'full_path': dir_path } } ) )
 
                 for filename in filenames:
                     file_path = os.path.relpath( os.path.join( dirpath, filename ), directory )
-                    file_path_hash = hashlib.sha1( file_path.encode('utf-8') ).hexdigest()
+                    file_path_hash = hashlib.sha1(file_path).hexdigest()
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: The 2 previous `encode()` calls were added in commit 040118aae35da2c72fbf6332b32e8f3e0025a341 to fix #673.

How can I reproduce the problem you are trying to fix here?

Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: On the face of it, that seems to be the exact same issue I was getting _until_ I removed the extra encoding.  My testing configuration is: 

```
[yoplait@malleus:~/work/import] cat ~/work/galaxy/config/galaxy.ini | grep import_dir
library_import_dir = /Users/yoplait/work/import

[yoplait@malleus:~/work/import] ls -l ~/work/import
total 0
drwxr-xr-x  3 yoplait  staff  102 Mar 10 19:23 Ăfolder

[yoplait@malleus:~/work/import] ls -l ~/work/import/Ăfolder 
total 8
-rw-r--r--  1 yoplait  staff  17 Mar 10 19:20 Ăfile.txt
```

Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: Just confirmed again, when I swap back over to 16.01 this API call breaks with the above configuration, whereas it passes on this branch:

```
http://192.168.0.101:8080/api/remote_files?target=importdir&format=jstree
```

Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: @dannon It seems that for some reason when `target=importdir`, `file_path` is of type `str`, while when `target=userdir` it is of type `unicode`.
The patch in https://gist.github.com/nsoranzo/c2e1878d4c5b8021dbf6 should fix both your issue and the one in #673.

Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: @nsoranzo Thanks, I'll track back and find out why that difference exists.  Might be reasonable to patch this here, might be much more reasonable to fix it at the source ;)  Either way, I'll follow up and can make the changes in that patch if it looks like the right way to do it.

Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: @dannon After a bit more investigations I've found out that the reason for the difference is that when `target=userdir`, `full_import_dir` is created by joining `user_login` (which is unicode) with `user_base_dir` (str). This obviously does not happen when `target=importdir`.

Also I can confirm that the version in my gist is the only one working on both str and unicode and in both Python 2 and 3.

Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: @nsoranzo I'm not a huge fan of unicodify(stuff).encode(); seems overzealous casting back and forth.  I've handled this differently by addressing the encoding earlier, which works for both (and future) code paths without additional unnecessary casting.  Let me know what you think; seems cleaner to me.

Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: @dannon As mentioned above, your solution will not work on Python 3, where all strings are unicode:

```
$ python3
>>> import hashlib
>>> s = 'Ăfolder'
>>> hashlib.sha1(s).hexdigest()
Traceback (most recent call last):
TypeError: Unicode-objects must be encoded before hashing
```

Like it or not, my solution is the only future proof (future being Python 2.7 & 3 support).

Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: Feels like a dirty hack, but whatever.  I feel like there are probably more important things to worry about than python3 at the moment, but I'll do it.

Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: Thanks @dannon, I just thought it was better to fix it "once and for all".

Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/remote_files.py]: It's just quite cumbersome to ask 'does this work in python3' for every change, when a ton of the app doesn't, the dependencies don't, and it's a really long way away.  (and all this, for a PR marked WIP and not requesting review yet)

**Comment by martenson** [comment]
given this is release bugfix can we please move this forward? (out of WIP)
thank you @dannon

**Comment by dannon** [comment]
@martenson Fine with me, I'll open another PR for the actual bug this is trying to address since neither of these spots were it.

**Comment by nsoranzo** [comment]
:+1:

**Comment by martenson** [comment]
Thank you @dannon @nsoranzo

**Pull request merged**